blockwork/build/interface.py

```diff
@@ -26,54 +29,46 @@
 
 class InterfaceError(Exception): ...
 
-class InterfaceDirection(Enum):
-    Input = auto()
-    Output = auto()
-
 # The resolved value type
 _RVALUE = TypeVar("_RVALUE")
 
-class Interface(Generic[_RVALUE]):
-    transform: "Transform"
-    name: str
-    direction: InterfaceDirection
-    connections: list[Self]
+@InitHooks()
+class Interface(Generic[_RVALUE], metaclass=keyed_singleton(inst_key=lambda i: (i.__class__, i.key()))):
+    input_transform: Optional["Transform"]
+    output_transforms: list["Transform"]
 
-    def __init__(self):
-        raise NotImplementedError
+    class Direction(Enum):
+        Input = auto()
+        Output = auto()
```

Comment:
Nested class? I realise this effectively creates a namespace, but that could be achieved by calling this `InterfaceDirection` instead

Reply:
It was previously done like that, but it was irritating when implementing `resolve_container` for interface subclasses that you had to import this other class to compare the direction to (especially if subclassing a subclass).

What I'll do is hedge and stick is_input/is_output properties on the direction and then subclasses can just use those